### PR TITLE
[Linting] Add pre-commit

### PR DESCRIPTION
**Pull request by jond01**
Resolves [ML-5290](https://jira.iguazeng.com/browse/ML-5290).

https://pre-commit.com/

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -595,25 +595,9 @@ html-docs-dockerized: build-test ## Build html docs dockerized
...
 .PHONY: lint
-lint: fmt-check lint-imports ## Run lint on the code
-
-.PHONY: fmt-check
-fmt-check: ## Check the code (using ruff)
-	@echo "Running ruff checks..."
-	python -m ruff check --exit-non-zero-on-fix
-	python -m ruff format --check
+lint:
+	pre-commit run --all-files
```
Comment (gtopper) [inline Makefile:600]: Basically what this does is equivalent to lint then fmt. If the format is wrong, a subsequent run will pass because the first run already formatted the code.

That's probably fine, but I thought it useful to point out the change in behaviour.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2023 Iguazio
```
Comment (gtopper) [inline .pre-commit-config.yaml:1]:
```suggestion
# Copyright 2024 Iguazio
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -36,19 +36,18 @@ jobs:
...
+    - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
         cache: pip
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip~=23.2.1
-        pip install -r dev-requirements.txt
+        pip install pre-commit==3.6.0
```
Comment (gtopper) [inline .github/workflows/ci.yaml:50]: Shouldn't this match the requirement in `dev-requirements.txt`?
Either `pre-commit~=3.6.0` or `pre-commit~=3.6` would probably be in order.

Comment (gtopper) [reply .github/workflows/ci.yaml:50]: I wonder whether the linter really doesn't need the requirements installed to properly validate (e.g. to validate correct number of arguments is passed to a function imported from a module in required library).

File: pyproject.toml [hunk 4]
```diff
@@ -5,10 +5,6 @@ select = [
     "E", # pycodestyle
     "I", # isort
 ]
-extend-exclude = [
-    "server/api/proto",
-    "playground",
-]
```
Comment (gtopper) [inline pyproject.toml:11]: Why is this deleted?

File: .pre-commit-config.yaml (hunk 2 above)
Comment (jond01) [reply .pre-commit-config.yaml:1]: This is the suggested template:
https://github.com/mlrun/mlrun/blob/ba399e9f487b769324d9f9e98a8a047ea9c39a74/.github/workflows/ci.yaml#L68

And I think that it should be changed for all the files 🤔
I'll try updating it here, but it's a wider issue

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (jond01) [reply .github/workflows/ci.yaml:50]: > Shouldn't this match the requirement in `dev-requirements.txt`? Either `pre-commit~=3.6.0` or `pre-commit~=3.6` would probably be in order.

I preferred stability in the CI and therefore specified `==`. I can change it to `~=`, but then it's less predictable. Let me know what you think.



> I wonder whether the linter really doesn't need the requirements installed to properly validate (e.g. to validate correct number of arguments is passed to a function imported from a module in required library).

"the linter" - what do you mean by that? pre-commit? Ruff?
The hooks are not local, and pre-commit manages the environments for each hook separately.
Please clarify this point with an explicit example.

File: pyproject.toml (hunk 4 above)
Comment (jond01) [reply pyproject.toml:11]: It wasn't needed in the first place. It's even more redundant now - as pre-commit runs only on tracked files, so there's no need to ignore untracked folders explicitly.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (jond01) [reply .pre-commit-config.yaml]: Ok, it worked 🙂
96bb5d0

File: Makefile (hunk 1 above)
Comment (jond01) [reply Makefile:600]: Right, good point. I'll write it in the message.
One may choose a Ruff integration for their IDE and catch this even earlier. I recommend Ruff's VSCode extension for VSCode users.

[the remark this answers is not in the document]
Comment (jond01) [reply Makefile]: > missing target note, 

Added in a4b959a.

> also, will that fail if a fix was applied? we need to in order to fail CI

In that case, pre-commit fails -> the step fails -> the CI linting job fails.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (gtopper) [reply .pre-commit-config.yaml]: It's the year the file was added. You don't need to update all the existing files.

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (gtopper) [reply .github/workflows/ci.yaml:50]: > "the linter" - what do you mean by that? pre-commit? Ruff?

Ruff.

> The hooks are not local, and pre-commit manages the environments for each hook separately.

Not sure what this means or how it applies here.

> Please clarify this point with an explicit example.

I did, in parentheses. Doesn't ruff need to have the dependencies installed to know if a function is called with valid arguments?

Comment (gtopper) [reply .github/workflows/ci.yaml:50]: Ok, never mind, it just doesn't take the arguments passed at all...

Comment (jond01) [reply .github/workflows/ci.yaml:50]: > Ok, never mind, it just doesn't take the arguments passed at all...

Right. There are other tools that check that, for example: mypy ([ML-5304](https://jira.iguazeng.com/browse/ML-5304)). Pylint also does some static analysis checks, but that part is not in Ruff AFAIK.

**Review by gtopper — APPROVED** [review]
Very nice! ✅

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (jond01) [reply Makefile]: > > also, will that fail if a fix was applied? we need to in order to fail CI
> 
> In that case, pre-commit fails -> the step fails -> the CI linting job fails.

See https://github.com/mlrun/mlrun/pull/4951/commits/d0ff837ce77d017c74a96baa56ee2819a74d8df7.

**Comment by jond01** [comment]
We currently do not plan to add pre-commit. It can serve as a future reference though in case pre-commit is added.
